Sympfy_Hamiltonian Analytic Expresser_Collins Kariuki

Collins-kariuk opened the pull request:
I made the necessary changes to the sympyfy function according to the recommendations by Alex Hampel.

The sympyfy function outputs a reduced analytical or numerical Hamiltonian in momentum space utilizing the Fourier series. The input of this function is a TRIQS tight-binding model that one can get from converting their model from either a Wannier90 model or PythTB model using TRIQS's handy function that enables said conversion.

This function has been tested on 2 compounds with different symmetries. These compounds include a Strontium compound (SrCaCu2O4) and a Palladium compound (PdCoO2). This function mainly follows from the equation in the attached screenshot. This function will help anyone, especially experimentalists, better understand their results in the tight-binding model landscape by visualizing a reduced analytical expression (and a numerical expression should they wish).

![Tight-Binding Hamiltonian in Momentum-Space from Real Space using the Fourier Series](https://github.com/TRIQS/triqs/assets/74094734/633b4fd8-8898-40ac-9fd7-204dcdfd4345)

Collins-kariuk committed:
Added the sympyfy function alongside its necessary dependencies

Added the sympyfy function into the utils.py file. In a nutshell, this function returns the analytical form of the momentum space hamiltonian of the tight-binding model from a tight-binding lattice object, by utlizing Fourier series.

Collins-kariuk committed:
Included some of Alex Hampel's recommendations. See the description:

- Updated where I called the import functions. I placed the sympy importations inside the sympfy function.
- updated the internal functions' names.
- updated the dosctring to include only 1 return value

Collins-kariuk committed:
included last recommendations from Alex Hampel

Collins-kariuk committed:
Updated docstring definition & included precision parameter

Collins-kariuk committed:
Deleted the useless comments before the sympyfy function

Collins-kariuk committed:
Create settings.json

Wentzell commented:
@Collins-kariuk Thank you for this useful addition! Please remove the `settings.json` file from your PR

Wentzell commented:
Could you add some unit tests for your function to the PR?

Collins-kariuk committed:
changed the name of sympyfy to TB_to_sympy and succinctly edited docstring and comments

Because it sounds a lot like sympify which is a core function in SymPy's library

Collins-kariuk committed:
changed TB_to_sympy method to handle only 2D systems

This is because, in the unit tests, the set up involves a simple 2D system. Instead of changing the unit tests to handle a simple 3D system, I changed my function to handle a 2D system. This did not require a lot of work on my end. Perhaps at a later date, we can incorporate tests for 3D systems.

Collins-kariuk committed:
included a test for the TB_to_sympy_2D

I made sure to test it on the numerical and analytical outputs. Since the system in the setup was very simple, I did not require to make multiple assertions/tests.

the-hampel commented on test/python/lattice_tight_binding.py:
```diff
@@ -69,6 +74,35 @@ def test_TBLattice(self):
         self.assertEqual(tbl, tbl_read)
+    
+    def test_TB_to_sympy_2D(self):
+        import sympy as sp
```
We should maybe make this test optional to not depend on sympy? We could add something like we did in dft_tools:
```
        try:
            import sympy as sp
            from sympy import cos
        except ImportError:
            print('ImportError: Sympy needs to be installed to run this test)
            try:
                mpi.MPI.COMM_WORLD.Abort(1)
            except:
                sys.exit()
```

Collins-kariuk replied:
Thank you for this recommendation. I appreciate it. It is fixed now.

the-hampel commented on test/python/lattice_tight_binding.py:
```diff
@@ -16,11 +16,16 @@
 from h5 import HDFArchive
 
 import numpy as np
+sys.path.append('/triqs-sympyfy/python/triqs/lattice/')
```
I think this is specific to your triqs install right? Please remove

Collins-kariuk replied:
The reason why I used `sys.path` was that I wanted to import the `utils.py` file into the unit test so that I could use `TB_to_sympy_2D`. After looking up how to do so, I found that I could use an absolute path to the file using `sys.path`. At the time, I did not think that would pose a problem since utils.py exists in the same folder on your (TRIQS's) end too, so it would work for both of us.
If there is another way I could import `utils.py`, I would appreciate your help in figuring this out.

Collins-kariuk replied:
I removed it.

the-hampel committed:
Delete settings.json

the-hampel reviewed (commented):
Thank you @Collins-kariuk for adding all the requested changes. I removed the settings.json file. Could you consider please my last two comments for the test? Thank you. Afterwards this can be merged from my side.

Collins-kariuk committed:
Updated the way I import sympy by incorporating a try-except block

Collins-kariuk committed:
Merge branch 'unstable_collins_sympy' of github.com:Collins-kariuk/triqs-sympyfy into unstable_collins_sympy

Collins-kariuk committed:
Added DS_Store to gitignore

Collins-kariuk committed:
deleted the sys.append absolute path file finder

Collins-kariuk committed:
Updated to include both 3D and 2D cases

Collins-kariuk committed:
Included test for new altered 2D + 3D function

the-hampel commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 doc/documentation/python_api
 doc/documentation/cpp_api
+.DS_Store
```
please add system specific files to be ignored in your global git config: https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/

Collins-kariuk replied:
Done.

Collins-kariuk committed:
updated TB_to_sympy with proper comments when handling 2D case

the-hampel committed:
fix sympy import

the-hampel committed:
fix try except import

the-hampel committed:
remove ds store from git ignore

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,171 @@ def TB_from_pythTB(ptb):
+        triqs tight binding object
+    analytical: boolean, default = True
+        a boolean which will cause the function will return an analytical Hamiltonian, when true, and 
+        an numerical Hamiltonian otherwise
```
```suggestion
        whether to return the Hamiltonian in analytical (true) or numerical (false) form.
```

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,171 @@ def TB_from_pythTB(ptb):
+    precision: integer, default = 6
+        specifies the precision of the floating point amplitudes. The default precision is 6 but the user
+        can decrease this parameter for better recognition of similar hopping amplitudes, particularly
+        for symmetrical hoppings across the crystal lattice
```
```suggestion
    precision: integer, default = 6
        specifies the number of digits in the floating point amplitudes. The default value is 6 but the user
        can decrease it to help recognize similar hopping amplitudes, particularly for symmetrical hoppings
        across the crystal lattice
```

Wentzell commented on test/python/lattice_tight_binding.py:
```diff
@@ -69,6 +70,29 @@ def test_TBLattice(self):
+    def test_TB_to_sympy_2D(self):
+        try:
+            import sympy as sp
+            from sympy import cos
```
Please make sure to include sympy in the CI environment

Collins-kariuk replied:
Fixed

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,171 @@ def TB_from_pythTB(ptb):
+    # convert to SymPy matrix to use substitutions method available in SymPy
+    Hk_numerical = sp.Matrix(Hk)
+
+    # obtaining individual displacement vectors
+    a1 = np.around(TBL_units[0], precision)
+    a2 = np.around(TBL_units[1], precision)
+    a3 = np.around(TBL_units[2], precision)
```
```suggestion
    # obtaining individual displacement vectors
    a1, a2, a3 = np.around(TBL_units, precision)
```

Also do you want to set precision already here or for the actual values `a1X_numerical` ?

Collins-kariuk replied:
I want to set `precision` at `aX` because I can't set `precision` later in the code. I cannot set `precision `at `a1X_numerical` because its type (`<class 'sympy.core.add.Add'>`) does not allow it to be rounded down by `precision`.

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,171 @@ def TB_from_pythTB(ptb):
+
+    # numerical dot products between unit vectors
+    # and momentum space matrix
+    a1k_numerical = a1.dot(k_space_matrix)[0]
+    a2k_numerical = a2.dot(k_space_matrix)[0]
+    a3k_numerical = a3.dot(k_space_matrix)[0]
```
Why not just
```suggestion
    a1k_n, a2k_n, a3k_n = np.around(TBL_units.dot(k_space_matrix), precision)
```
You could drop the part above.

Collins-kariuk replied:
I cannot set `precision` at `a1X_numerical` because its type (`<class 'sympy.core.add.Add'>`) does not allow it to be rounded down by `precision`.

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,171 @@ def TB_from_pythTB(ptb):
+    Hk_numerical = Hk_numerical.subs(a2k, a2k_numerical)
+    Hk_numerical = Hk_numerical.subs(a3k, a3k_numerical)
+
+    # converting numerical Hamiltonian to NumPy array
+    Hk_numerical = np.array(Hk_numerical)
```
```suggestion
    return np.array(Hk_numerical)
```

Collins-kariuk replied:
I would like to first check whether the user wants to return an analytical or numerical form before returning anything. Moreover, returning `np.array(Hk_numerical)` that early on will cause the function to "exit" before even initiating the checks afterward.

Collins-kariuk replied:
Since the checks I instituted, `_has_complex_exponential_sympy(matrix)` and `_is_hermitian_sympy(matrix)`, depend on `Hk_numerical` being a NumPy array, I first convert it from a SymPy matrix to a NumPy array and then do the checks.

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,171 @@ def TB_from_pythTB(ptb):
+                                Choosing a different unit cell could make 
+                                your Hamiltonian expression real.""")
+    
+    # returning analytical or numerical form of Hamiltonian
+    if analytical:
+        return Hk
+    return Hk_numerical
```
```suggestion
```

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,171 @@ def TB_from_pythTB(ptb):
+    
+    # rewriting exponential terms in Hamiltonian expression in terms of cosine
+    for i, j in itp(range(num_orb), repeat = 2):
+        Hk[i, j] = Hk[i, j].rewrite(sp.cos)
```
```suggestion
        Hk[i, j] = Hk[i, j].rewrite(sp.cos)
        
        if analytical: return Hk
```

Wentzell reviewed (changes requested):
Thank you @Collins-kariuk for this addition
See comments below

Collins-kariuk committed:
Rephrased doctoring description of 'analytical' parameter

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Rephrased doctstring description of 'precision' parameter

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Added sympy as a required library in CI

Collins-kariuk committed:
Merge branch 'unstable_collins_sympy' of github.com:Collins-kariuk/triqs-sympyfy into unstable_collins_sympy

Collins-kariuk committed:
Condensed how I get the individual displacement vectors into a one-liner

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Condensed how I get the numerical dot products into a one-liner

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Shortened variable names of numerical dot products

Collins-kariuk committed:
Removed the check at the end where I check to see whether the user want an analytical or numerical Ham

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Check whether user wants an analytical Ham form right after rewriting Hk in terms of cos

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Changed indentation of `analytical` check

Collins-kariuk committed:
Removed precision around numerical dot products because it results in error

Collins-kariuk committed:
Changed how return `Hk_numerical`

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Changed positions of the checks on the Hamiltonian

Collins-kariuk commented:
@Wentzell Thank you for your recommendations. I have implemented them. Please let me know if there is anything else I can do.

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,168 @@ def TB_from_pythTB(ptb):
+    Hk_numerical = sp.Matrix(Hk)
+
+    # obtaining individual displacement vectors
+    a1, a2, a3 = np.around(TBL_units, precision)
```
```suggestion
    TBL_units_prec = np.around(TBL_units, precision)
```

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,168 @@ def TB_from_pythTB(ptb):
+    # obtaining individual displacement vectors
+    a1, a2, a3 = np.around(TBL_units, precision)
+
+    # numerical dot products between unit vectors
+    # and momentum space matrix
+    a1k_n = a1.dot(k_space_matrix)[0]
+    a2k_n = a2.dot(k_space_matrix)[0]
+    a3k_n = a3.dot(k_space_matrix)[0]
```
```suggestion
    # dot product between unit vectors and momentum vector
    k_vec = sp.symbols("kx ky kz", real = True)
    a1k_n, a2k_n, a3k_n = TBL_units_prec.dot(k_vec)
```

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,168 @@ def TB_from_pythTB(ptb):
+    a2k_n = a2.dot(k_space_matrix)[0]
+    a3k_n = a3.dot(k_space_matrix)[0]
+    
+    # performing numerical dot product substitutions
+    Hk_numerical = Hk_numerical.subs(a1k, a1k_n)
+    Hk_numerical = Hk_numerical.subs(a2k, a2k_n)
+    Hk_numerical = Hk_numerical.subs(a3k, a3k_n)
```
```suggestion
    # substitute numerical unit vectors into H_k
    Hk_numerical = Hk_numerical.subs([(a1k, a1k_n), (a2k, a2k_n), (a3k, a3k_n)])
```

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,168 @@ def TB_from_pythTB(ptb):
+    Hk_numerical = Hk_numerical.subs(a2k, a2k_n)
+    Hk_numerical = Hk_numerical.subs(a3k, a3k_n)
+
+    # converting numerical Hamiltonian to NumPy array
+    Hk_numerical = np.array(Hk_numerical)
```
The comment is a bit redundant?

Collins-kariuk replied:
I agree. I deleted the comment.

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,168 @@ def TB_from_pythTB(ptb):
+    # imaginary number
+    I = sp.I
+
+    # matrix from axis directions in momentum space
+    kx, ky, kz = sp.symbols("kx ky kz", real = True)
+    k_space_matrix = sp.Matrix([kx, ky, kz])
+
```
```suggestion
```

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,168 @@ def TB_from_pythTB(ptb):
+        return True
+    
+    # performing the check on the analytical Hamiltonian
+    if not _is_hermitian_sympy(Hk): return warnings.warn("The resulting Hamiltonian is not hermitian.")
+    if _has_complex_exponential_sympy(Hk): return warnings.warn("""Your expression has a complex exponential. 
+                                                                    Choosing a different unit cell could make 
+                                                                    your Hamiltonian expression real.""")
```
```suggestion
    if not _is_hermitian_sympy(Hk): warnings.warn("The resulting Hamiltonian is not hermitian.")
    if _has_complex_exponential_sympy(Hk): warnings.warn("""Your expression has a complex exponential. 
                                                                    Choosing a different unit cell could make 
                                                                    your Hamiltonian expression real.""")
```

Wentzell commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +248,168 @@ def TB_from_pythTB(ptb):
+    # converting numerical Hamiltonian to NumPy array
+    Hk_numerical = np.array(Hk_numerical)
+
+    if not _is_hermitian_sympy(Hk_numerical): return warnings.warn("The resulting Hamiltonian is not hermitian.")
+    if _has_complex_exponential_sympy(Hk_numerical): return warnings.warn("""Your expression has a complex exponential. 
+                                                                            Choosing a different unit cell could make 
+                                                                            your Hamiltonian expression real.""")
```
```suggestion
    if not _is_hermitian_sympy(Hk_numerical): return warnings.warn("The resulting Hamiltonian is not hermitian.")
    if _has_complex_exponential_sympy(Hk_numerical): return warnings.warn("""Your expression has a complex exponential. 
                                                                            Choosing a different unit cell could make 
                                                                            your Hamiltonian expression real.""")
```
```suggestion
    if not _is_hermitian_sympy(Hk_numerical): warnings.warn("The resulting Hamiltonian is not hermitian.")
    if _has_complex_exponential_sympy(Hk_numerical): warnings.warn("""Your expression has a complex exponential. 
                                                                            Choosing a different unit cell could make 
                                                                            your Hamiltonian expression real.""")
```

Collins-kariuk committed:
Conglomerated all displacement vectors into a matrix whose elements have been specified by `precision`

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Conglomerated dot product between k-vec and displacement vectors

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Explicitly returned warnings

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Enhanced way in which numerical substitution is performed

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Moved location of where I define the k_space_matrix/coordinate axes

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk committed:
Removed redundant comment

Collins-kariuk committed:
Changed how I return the warnings

Co-authored-by: Nils Wentzell <Wentzell@users.noreply.github.com>

Collins-kariuk commented:
@Wentzell Thank you for the additional recommendations. I have implemented them. Again, if there is anything else I can do, please don't hesitate to inform me.

Wentzell reviewed (approved):
Thank you @Collins-kariuk for this valuable contribution!
